### PR TITLE
Added scripts for 2016-09-12 back-to-school mailing

### PR DESCRIPTION
**Pull request by aoby**
Preliminary review for Monday's back to school mailing. Some of the requirements may still change.

### CODE REVIEW

File: bin/mail/2016-09-12-back-to-school/generate_list_functions.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  {}.tap do |results|
+    DASHBOARD_DB.fetch(query).each do |teacher|
+      email = teacher[:email]
+      results[email] = {email: email, name: teacher[:name], id: teacher[:id]}
+    end
+  end
+end
+
+def query_workshop_section_ids
+  DASHBOARD_DB.fetch("SELECT id FROM sections WHERE section_type LIKE '%workshop'").map(:id)
```
Comment (ashercodeorg) [inline bin/mail/2016-09-12-back-to-school/generate_list_functions.rb]: Is `map(:id)` proper syntax?

File: bin/mail/2016-09-12-back-to-school/generate_list_functions.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+# PD comes in 3 forms: old ops workshops, old K5 workshops, and new pd_workshops.
+# The old ops workshop attendance is tracked in teacher_attendance.
+# For the old K5 and all the new workshops, we can track teachers through section membership.
+def query_all_pd_teachers
+  ops_pd_teachers = query_dashboard_teachers <<-SQL
+    SELECT DISTINCT users.id, users.email, users.name
+    FROM users
+    JOIN workshop_attendance ON (workshop_attendance.teacher_id = users.id)
+    JOIN segments ON (segments.id = workshop_attendance.segment_id)
+    JOIN workshops ON (workshops.id = segments.workshop_id)
```
Comment (ashercodeorg) [inline bin/mail/2016-09-12-back-to-school/generate_list_functions.rb]: nit: Making the INNER JOIN explicit would be my preference, though this is (only?) a matter of personal style.

Comment (ashercodeorg) [inline bin/mail/2016-09-12-back-to-school/generate_list_functions.rb]: nit: Parens are unnecessary.

File: bin/mail/2016-09-12-back-to-school/generate_list_functions.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    WHERE users.user_type = 'teacher'
```
Comment (ashercodeorg) [inline bin/mail/2016-09-12-back-to-school/generate_list_functions.rb]: Is there a reason to have this here? Does it make sense to require `users.email` to be non-empty?

Also below.

File: bin/mail/2016-09-12-back-to-school/generate_list_functions.rb (hunk 1 above)
Comment (aoby) [reply bin/mail/2016-09-12-back-to-school/generate_list_functions.rb]: Yes, it's sequel's version of ActiveRecord's `pluck` to return an array of column values. See http://sequel.jeremyevans.net/rdoc/files/doc/querying_rdoc.html#label-As+an+Array+of+Column+Values

File: bin/mail/2016-09-12-back-to-school/generate_list_functions.rb (hunk 3 above)
Comment (aoby) [reply bin/mail/2016-09-12-back-to-school/generate_list_functions.rb]: I suppose we can assume they're teachers if they're in the workshops. I'll verify email instead: `WHERE users.email IS NOT NULL AND users.email <> ''`

File: bin/mail/2016-09-12-back-to-school/generate_list_functions.rb (hunk 2 above)
Comment (aoby) [reply bin/mail/2016-09-12-back-to-school/generate_list_functions.rb]: will fix

**Comment by aoby** [comment]
Update: Now we're doing 4 email templates instead of 3, splitting the Pd list into K5 and 6-12. PTAL.

File: bin/mail/2016-09-12-back-to-school/generate_list_functions.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env ruby
+require_relative '../mailing-common/mailing-list-utils'
+require 'active_support'
+require 'active_support/core_ext/object/blank'
+
+ALL_EMAILS = Set.new
```
Comment (ashercodeorg) [inline bin/mail/2016-09-12-back-to-school/generate_list_functions.rb]: This doesn't appear to be a constant, so should not be SCREAMING_CASE.

File: bin/mail/2016-09-12-back-to-school/generate_list_functions.rb [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+def dedupe(contacts)
+  contacts.except! *ALL_EMAILS
+  ALL_EMAILS.merge contacts.keys
```
Comment (ashercodeorg) [inline bin/mail/2016-09-12-back-to-school/generate_list_functions.rb]: This should be `merge!`?

**Comment by ashercodeorg** [comment]
LGTM, beyond new comments.

File: bin/mail/2016-09-12-back-to-school/generate_list_functions.rb (hunk 5 above)
Comment (aoby) [reply bin/mail/2016-09-12-back-to-school/generate_list_functions.rb]: You would think, but apparently not with [Set](http://ruby-doc.org/stdlib-2.3.1/libdoc/set/rdoc/Set.html#method-i-merge)

File: bin/mail/2016-09-12-back-to-school/generate_list_functions.rb (hunk 4 above)
Comment (aoby) [reply bin/mail/2016-09-12-back-to-school/generate_list_functions.rb]: Will change to `@all_emails`

**Comment by aoby** [comment]
Another requirement change: removing HOC organizers from the lists.

**Pull request merged**